// openapi-server/dist/routes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = __importDefault(require("express"));
const openApiVersion_1 = __importDefault(require("./lib/openApiVersion"));
const openApiVersion_2 = __importDefault(require("./types/openApiVersion"));
const v2_1 = __importDefault(require("./routes/v2"));
const v3_1 = __importDefault(require("./routes/v3"));
exports.operations = {
    get: 'get',
    put: 'put',
    post: 'post',
    del: 'delete',
    delete: 'delete',
    options: 'options',
    head: 'head',
    patch: 'patch'
};
exports.default = (spec) => {
    const router = express_1.default.Router();
    const version = openApiVersion_1.default(spec);
    const paths = spec.paths;
    switch (version) {
        case openApiVersion_2.default.v2:
            return v2_1.default(router, paths);
        case openApiVersion_2.default.v3:
            return v3_1.default(router, paths);
    }
    throw new Error('Unsupported openApi version');
};
